shared/src/ledger/protocol/transactions/votes/update.rs

```diff
@@ -0,0 +1,547 @@
+                nonce: 0.into(),
+                transfers: vec![],
+            }
+        }
+
+        /// Writes an initial [`Tally`] to storage, based on the passed `votes`.
+        pub(super) fn setup_tally(
+            storage: &mut TestStorage,
+            event: &EthereumEvent,
+            keys: &vote_tallies::Keys<EthereumEvent>,
+            votes: HashSet<(Address, BlockHeight, FractionalVotingPower)>,
+        ) -> Result<Tally> {
+            let voting_power: FractionalVotingPower =
+                votes.iter().cloned().map(|(_, _, v)| v).sum();
+            let tally = Tally {
+                voting_power: voting_power.to_owned(),
+                seen_by: votes.into_iter().map(|(a, h, _)| (a, h)).collect(),
+                seen: voting_power > FractionalVotingPower::TWO_THIRDS,
+            };
+            votes::storage::write(storage, keys, event, &tally)?;
+            Ok(tally)
+        }
+    }
+
+    #[test]
+    fn test_vote_info_new_empty() -> Result<()> {
+        let voting_powers = HashMap::default();
+
+        let vote_info = NewVotes::new(Votes::default(), &voting_powers)?;
+
+        assert!(vote_info.voters().is_empty());
+        assert_eq!(vote_info.into_iter().count(), 0);
+        Ok(())
+    }
+
+    #[test]
+    fn test_vote_info_new_single_voter() -> Result<()> {
+        let validator = address::testing::established_address_1;
+        let vote_height = || BlockHeight(100);
+        let voting_power = || FractionalVotingPower::new(1, 3).unwrap();
+        let vote = || (validator(), vote_height());
+        let votes = Votes::from([vote()]);
+        let voting_powers = HashMap::from([(vote(), voting_power())]);
+
+        let vote_info = NewVotes::new(votes, &voting_powers)?;
+
+        assert_eq!(vote_info.voters(), BTreeSet::from([validator()]));
+        let votes: BTreeSet<_> = vote_info.into_iter().collect();
+        assert_eq!(
+            votes,
+            BTreeSet::from([(validator(), vote_height(), voting_power())]),
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_vote_info_new_error() -> Result<()> {
+        let validator = address::testing::established_address_1;
+        let vote_height = || BlockHeight(100);
+        let vote = || (validator(), vote_height());
+        let votes = Votes::from([vote()]);
+        // voting powers map is missing vote
+        let voting_powers = HashMap::default();
+
+        let result = NewVotes::new(votes, &voting_powers);
+
+        assert!(result.is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn test_vote_info_without_voters() -> Result<()> {
+        let validator = address::testing::established_address_1;
+        let vote_height = || BlockHeight(100);
+        let voting_power = || FractionalVotingPower::new(1, 3).unwrap();
+        let vote = || (validator(), vote_height());
+        let votes = Votes::from([vote()]);
+        let voting_powers = HashMap::from([(vote(), voting_power())]);
```

Comment:
again, what's the idea with the closures?

Reply:
it was to save on having to `.clone()` in all places but the last occurence of these variables otherwise, but will change it to the other way as that is more consistent with the rest of the code